### PR TITLE
ARXIVNG-2480 don't store task state in s3; use Redis instead

### PR DESCRIPTION
**Pull request by erickpeirson**
From ``DECISIONS.md``:

## 2019-07-23: Don't store task state in S3

In previous versions, both the compilation product (e.g. PDF) and information about the state of the compilation task were stored in S3. The main driver for this was to avoid introducing a new state store (e.g. a database) just to store a little bit of metadata about a compilation task.

As reported by e.g. https://github.com/arXiv/arxiv-submission-ui/issues/123 we ran into some performance issues, and some lost state. Specifically, the state of the compilation process was not always successfully stored in S3, leading to an unacceptable split-brain situation. The fix for this seemed to be to consult the task backend (Redis) for the "real" state of the compilation task.... which begs the question of why we would store compilation status metadata anywhere else in the first place.

A simpler approach (that still avoids a database) is to rely entirely on the task/result backend (Redis) for the state of the compilation process. By setting [``result_extended =
True``](http://docs.celeryproject.org/en/latest/userguide/configuration.html#result-extended) and handling failures gracefully, we can ensure that the original parameters for the compilation (e.g. owner) can always be successfully obtained.

Incidentally, this also speeds up read-only requests for compilation status by something like 10x.

### CODE REVIEW

**Comment by erickpeirson** [comment]
@DavidLFielding I still think that your timeout suggestion in arXiv/arxiv-submission-ui#123 is a good one. Wanted to start with this change, because I believe that it will eliminate 99.9% of the cases in which a timeout would have been required (since the root cause of the problem you're seeing is the split brain described above).

**Comment by DavidLFielding** [comment]
One question I have concerns the original rationale for storing the compilation state in S3. Was the original intent to provide a check (staleness) for files (PDF) stored in S3? That is if we have a PDF and a more recent compilation status object, do we make decisions based on these two items being available for inspection? So if the compilation status is more recent than the PDF do we indicate that the PDF may be stale or out of date? Or possibly invalidate/expire the PDF?

This leads to another question. What is the 'status' of an existing PDF when a request to recompile the paper's source has been received by the compilation service? I would expect that the PDF would be 'stale' or 'invalid' in the event the paper source checksum has changed and the paper is getting recompiled by the compilation service. It should also be possible to recompile without a change in checksum since fixes to TeX might warrant a bulk recompilation of papers.

So in some ways, I'm asking if there is a potential loss of originally intended functionality (future).

**Comment by erickpeirson** [comment]
> Was the original intent to provide a check (staleness) for files (PDF) stored in S3? That is if we have a PDF and a more recent compilation status object, do we make decisions based on these two items being available for inspection? So if the compilation status is more recent than the PDF do we indicate that the PDF may be stale or out of date? Or possibly invalidate/expire the PDF?

A staleness check was not a driver that went into the S3 decision. The (bad) assumption was that the result of a second compilation attempt on the same input files would have precisely the same output, so the case that you're describing really wasn't considered.

> This leads to another question. What is the 'status' of an existing PDF when a request to recompile the paper's source has been received by the compilation service? I would expect that the PDF would be 'stale' or 'invalid' in the event the paper source checksum has changed and the paper is getting recompiled by the compilation service. It should also be possible to recompile without a change in checksum since fixes to TeX might warrant a bulk recompilation of papers.

Indeed, we do support a ``force`` parameter for recompilation, so we **should** consider that case. It would be easy to poll Redis for the latest task state, since we determine the task ID deterministically from the source package details (id, checksum) and output format.

I will create a ticket so that we don't lose track of this.

**Comment by DavidLFielding** [comment]
You do not indicate how you detected the S3 access failures so I'm assuming such errors are ending up in a log. Are these errors recoverable? Is there the potential of the same error occurring when writing the PDF or other data written to S3 elsewhere? In this case, I'm wondering if we need to build in some type of retry/recovery into interactions with external services like S3. This depends on your analysis of the errant condition.

I suspect there may be timing issues here with rapid changes in the state with consumers trying to read the state. I also expect keeping the state locally in redis will definitely speed things up.

I'm going to take another quick pass over the code and wrap up my review. I'm thinking to do a more intensive exercise of the development server after the several pending PRs are merged. Let me know if you'd prefer I continued on this PR. There are some other large ones waiting.

**Comment by erickpeirson** [comment]
> Are these errors recoverable? Is there the potential of the same error occurring when writing the PDF or other data written to S3 elsewhere? In this case, I'm wondering if we need to build in some type of retry/recovery into interactions with external services like S3. This depends on your analysis of the errant condition.

Yes, this is an excellent point. Indeed, most of these failures are recoverable. One of the most common cases that I see is that the app gets a fresh AWS key-pair authorized for writing to S3, and a request that requires a write to S3 comes in within a second or two of the new credential. Since AWS credentials are not immediately consistent, this can lead to an authorization failure. Simply retrying the request will almost always resolve the situation.

We would want to implement the retry here: https://github.com/arXiv/arxiv-compiler/blob/c1166b629f607182f9d012e19bddb77d33d72d7f/compiler/services/store/__init__.py#L248-L254

**Comment by erickpeirson** [comment]
> I'm going to take another quick pass over the code and wrap up my review. I'm thinking to do a more intensive exercise of the development server after the several pending PRs are merged. Let me know if you'd prefer I continued on this PR. There are some other large ones waiting.

Good plan. This sounds right to me.

**Comment by DavidLFielding** [comment]
A quick follow-up on the fresh AWS key-pair authorized for writing to S3 causing failures. I'm wondering if the code that writes to S3 could recognize the app's receipt of a new key-pair and either wait a set amount of time, make some S3 request until authorization succeeds, or adjust to retry with a special 'key-pair' transition exception that we expect to disappear and can safely ignore.

While I suspect these key-pair updates should not occur too frequently I was able to send the compilation service off into oblivion on a regular basis. This might hint at additional issues unless the key-pair lifetime is set to an extremely short period. Back to code review.

File: compiler/compiler.py [hunk 1]
```diff
@@ -194,23 +191,24 @@ def get_task(src_id: str, chk: str, fmt: Format = Format.PDF) -> Task:
     task_id = _get_task_id(src_id, chk, fmt)
     result = do_compile.AsyncResult(task_id)
     stat = Status.IN_PROGRESS
-    owner: Optional[str] = None
     reason = Reason.NONE
+    _info: Dict[str, str]
     if result.status == 'PENDING':
         raise NoSuchTask(f'No such task: {task_id}')
```
Comment (DavidLFielding) [inline compiler/compiler.py:200]: The PENDING state means different things but I'm surprised to see it triggers NoSuchTask exception. Is this like a failed launch attempt or halted state? Not really part of this PR. I suspect the meaning of this state evolved over time. Pending gives me the impression the job is waiting and will eventually run. No biggie.

Comment (erickpeirson) [reply compiler/compiler.py:200]: It's a foible of Celery. When you instantiate an ``AsyncResult``, it will wait lazily to actually retrieve anything from redis until you access its properties, and show a ``PENDING`` state (even if there is no such task). And in some cases (maybe this is fixed now), but you could end up just waiting+blocking on non-existant tasks.

So what I generally do is set the state to something other than ``PENDING`` when a task is created. ``SENT`` in this case, IIRC. That way if we see ``PENDING`` we know for sure that the task doesn't actually exist.

**Review by DavidLFielding — APPROVED** [review]
I'm able to follow the relocation of compilation state storage from S3 to Redis. Code looks fine. 

I decided to fire up the compiler using the instructions in the README. It's been a while so I could be doing something wrong.

The compilation of the example [1602.00123] from the README appears to work.

compiler-test-worker_1      | [2019-07-29 11:56:42,192: INFO/ForkPoolWorker-2] Task compiler.compiler.do_compile[1602.00123/IlRodSwgMDMgSmFuIDIwMTkgMDE6MDQ6MzMgR01UIg==/pdf] succeeded in 12.18705169999157s: {'source_id': '1602.00123', 'output_format': 'pdf', 'checksum': 'IlRodSwgMDMgSmFuIDIwMTkgMDE6MDQ6MzMgR01UIg==', 'task_id': '1602.00123/IlRodSwgMDMgSmFuIDIwMTkgMDE6MDQ6MzMgR01UIg==/pdf', 'status': 'completed', 'reason': None, 'description': 'Success!', 'size_bytes': 689037, 'owner': None}

When I quickly take the location output from the curl request and enter it into my browser I receive an error: I don't believe I should receive an error when following this sequence of events. Especially with how long it takes me to copy and paste the URL into my browser.  This observation may not be related to this PR.

curl -XPOST -i -H 'Authorization: <token>' -d '{"source_id":"1602.00123","checksum":"\"Thu, 03 Jan 2019 01:04:33 GMT\"","format":"pdf"}' http://localhost:8000/
HTTP/1.1 202 ACCEPTED
Content-Type: application/json
Content-Length: 3
Location: http://localhost:8000/1602.00123/IlRodSwgMDMgSmFuIDIwMTkgMDE6MDQ6MzMgR01UIg%3D%3D/pdf

{"reason":"The server encountered an internal error and was unable to complete your request. Either the server is overloaded or there is an error in the application."}

When I refresh I get the response I was expecting:

{"checksum":"IlRodSwgMDMgSmFuIDIwMTkgMDE6MDQ6MzMgR01UIg==","description":"","output_format":"pdf","owner":null,"reason":null,"size_bytes":0,"source_id":"1602.00123","status":"completed","task_id":"1602.00123/IlRodSwgMDMgSmFuIDIwMTkgMDE6MDQ6MzMgR01UIg==/pdf"}

Only ODD thing here is "size_bytes":0. This seems wrong to me especially when compiler log output shows:

'Success!', 'size_bytes': 689037,

I would also note that while experimenting I entered an incorrect checksum and it seemed I would also get a status of completed. That is, it was difficult to clearly detect an error. I may not be looking at the logs hard enough. 

I will go ahead and approve but do request you take a look to see if there is a minor issue with saving/retrieving state information.

**Comment by erickpeirson** [comment]
> When I quickly take the location output from the curl request and enter it into my browser I receive an error: I don't believe I should receive an error when following this sequence of events. Especially with how long it takes me to copy and paste the URL into my browser. This observation may not be related to this PR.

This was definitely a bug. Was not correctly handling the case that no metadata/result is set for the task. This is fixed in 997bcd3, along with the missing file size

**Comment by erickpeirson** [comment]
> I would also note that while experimenting I entered an incorrect checksum and it seemed I would also get a status of completed. That is, it was difficult to clearly detect an error. I may not be looking at the logs hard enough.

This looks like a bug. Will create a separate issue to address, as I think it needs focused attention.
